docs/examples/GradleExamples/GradleJava7Example/build.gradle

```diff
@@ -53,10 +44,7 @@ allprojects {
                 //'-Awarns',
                 "-Xbootclasspath/p:${configurations.checkerFrameworkAnnotatedJDK.asPath}"
         ]
-        if (targetJavaVersion.isJava7()) {
-            compile.options.compilerArgs += ['-source', '7', '-target', '7']
-            options.fork = true
-            options.forkOptions.jvmArgs += ["-Xbootclasspath/p:${configurations.checkerFrameworkJavac.asPath}"]
-        }
+        options.fork = true
+        options.forkOptions.jvmArgs += ["-Xbootclasspath/p:${configurations.checkerFrameworkJavac.asPath}"]
```

Comment:
The expected errors for this example are:
```
/Users/smillst/jsr308/checker-framework/docs/examples/GradleExamples/GradleJava7Example/src/main/java/org/checkerframework/example/GradleExample.java:33: error: [assignment.type.incompatible] incompatible types in assignment.
        /*@NonNull*/ Object nn = nullable; // error on this line
                                 ^
  found   : @Initialized @Nullable Object
  required: @UnknownInitialization @NonNull Object
/Users/smillst/jsr308/checker-framework/docs/examples/GradleExamples/GradleJava7Example/src/main/java/org/checkerframework/example/GradleExample.java:36: error: [argument.type.incompatible] incompatible types in argument.
        list.add(null); // error on this line
                 ^
  found   : null
  required: @Initialized @NonNull String
2 errors
:compileJava FAILED
```

Without this line the errors are:
```
/Users/smillst/jsr308/checker-framework/docs/examples/GradleExamples/GradleJava7Example/src/main/java/org/checkerframework/example/GradleExample.java:26: error: [assignment.type.incompatible] incompatible types in assignment.
    public static /*@Nullable*/ Object nullable = null;
                                                  ^
  found   : null
  required: @Initialized @NonNull Object
/Users/smillst/jsr308/checker-framework/docs/examples/GradleExamples/GradleJava7Example/src/main/java/org/checkerframework/example/GradleExample.java:36: error: [argument.type.incompatible] incompatible types in argument.
        list.add(null); // error on this line
                 ^
  found   : null
  required: @Initialized @NonNull String
2 errors
:compileJava FAILED
```
I don't know understand why this is the case.  @wmdietl any ideas?

Reply:
The first error message is when annotations-in-comments are correctly used; the second error messages ignore annotations-in-comments.

Is it somehow possible to put `${configurations.checkerFrameworkJavac.asPath}` on the normal classpath for the JVM that runs the compiler?
It shouldn't be necessary to use the bootclasspath, unless that is the only way to change which compiler Gradle uses.
@CharlesZ-Chen can you try this example?

Reply:
Hi @smillst @wmdietl,

I tried running this example by enter directory `docs/examples/GradleExamples/GradleJava7Example` and then run `gradle build`, but I got an strange exception from gradle:

```
java.lang.BootstrapMethodError: call site initialization exception
        at java.lang.invoke.CallSite.makeSite(CallSite.java:341)
        at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:307)
        at java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:297)
        at java.io.ObjectInputStream.<clinit>(ObjectInputStream.java:3578)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:104)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:62)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:62)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:67)
Caused by: java.lang.ClassCastException: bootstrap method failed to produce a CallSite
        at java.lang.invoke.CallSite.makeSite(CallSite.java:332)
        ... 7 more
```

I found this post through google, it seems someone tried use Checker Framework with gradle and encountered this problem before:
https://discuss.gradle.org/t/bootstrapmethoderror-with-custom-bootclasspath-in-android/21703

But there were no response to that post:-(

Any ideas on this exception?

Thanks!

Reply:
If I removed this line,

```
options.forkOptions.jvmArgs += ["-Xbootclasspath/p:${configurations.checkerFrameworkJavac.asPath}"]
```

then I got same result as @smillst  got (the case without putting jsr308 javac on bootclasspath).

Reply:
Do you get this exception only after the change? Does master work?

Reply:
master worked. Below is the result I got on master branch:

```
checker-framework/docs/examples/GradleExamples/GradleJava7Example/src/main/java/org/checkerframework/example/GradleExample.java:26: error: [assignment.type.incompatible] incompatible types in assignment.
    public static /*@Nullable*/ Object nullable = null;
                                                  ^
  found   : null
  required: @Initialized @NonNull Object
checker-framework/docs/examples/GradleExamples/GradleJava7Example/src/main/java/org/checkerframework/example/GradleExample.java:36: error: [argument.type.incompatible] incompatible types in argument.
        list.add(null); // error on this line
                 ^
  found   : null
  required: @Initialized @NonNull String
```

Reply:
Are you using a Java 8 JVM? The version in master doesn't put the 308 compiler on the BCP when on 8.
So it would be nice to find a way to use the 308 compiler on a 7 or 8 JVM, so that in either setting we get annotations-in-comments.

Reply:
I'm not sure about that error, but I used Gradle 3.5 to build the example.

Reply:
Hi Prof. Dietl @wmdietl , I investigated this issue a bit, and I think it should be able to set 308 javac on the classpath when starting the JVM for compilation, and I tried as following:

(Instead of setting javac on bootclasspath, set it to classpath)

```gradle
options.fork = true
options.forkOptions.jvmArgs += ["-classpath", "${configurations.checkerFrameworkJavac.asPath}"]
```

However, the gradle build result still kept same as before. It seems ignore this classpath setting...

I'm not sure if I set the classpath for gradle fork JVM correctly (by adding classpath in `jvmArgs`), but it seems this is the most possible way of doing that after I searching the gradle APIs doc.

Also, I'm not sure if my gradle/ OS env might has some influence on it, as I even cannot get the bootclasspath case working on my machine. (Even I switched to gradle 3.5, I still got the strange bootstrap Exception).

@smillst , could you please try setting `options.forkOptions.jvmArgs += ["-classpath", "${configurations.checkerFrameworkJavac.asPath}"]` on your machine to see if that works? Theoretically it should work as this seems is the same thing we do in `CheckerMain`.

Thanks!

Reply:
I tried that, but annotations in comments are still ignored.  I think we should just leave it on the bootclasspath.

Reply:
I'm fine with keeping jsr308-langtools on the bootclasspath to keep Gradle working. It is rather strange that it works without BCP in our other uses.

Reply:
One hope of removing it from the BCP was that https://github.com/typetools/checker-framework/issues/1459 would be fixed. Let's just use that issue for future Gradle work.

Reply:
The problem is that the compiler jar file contains java/lang classes, e.g.:

```
     3037  2017-08-03 09:06   com/sun/tools/javac/util/UnsharedNameTable.class
     2115  2017-08-03 09:06   com/sun/tools/javac/util/Warner.class
      439  2017-08-03 09:06   java/lang/FunctionalInterface.class
     1479  2017-08-03 09:06   java/lang/annotation/ElementType.class
      551  2017-08-03 09:06   java/lang/annotation/Repeatable.class
     1045  2017-08-03 09:06   java/lang/invoke/LambdaConversionException.class
     1463  2017-08-03 09:06   java/lang/invoke/LambdaMetafactory.class
     5068  2017-08-03 09:06   javax/annotation/processing/AbstractProcessor.class
      192  2017-08-03 09:06   javax/annotation/processing/Completion.class
```

We need these classes to run on a Java 7 JVM, because the compiler depends on these classes.
However, when we run on a Java 8 JVM, as we now must, this creates an inconsistency between the Java 8 classes in the rest of the JDK and these few classes that ship with our compiler, causing the strange exception.

I removed these extra classes in jsr308-langtools, so with the next release, we will be able to put our compiler on the bootclasspath and annotations-in-comments should work.
I'm not sure whether there is an easy way to test this before the release.

If we wanted to also support Java 7 execution, we would need to split out these files into a separate module and selectively include it. As we don't care about 7 any longer, we can drop this.
I'll push this change to this PR.

Reply:
https://github.com/typetools/checker-framework/pull/1487 also includes the required manual updates.

Reply:
Thanks for this explanation!